content/blog/2019-10-02-codestitcher.md

```diff
@@ -0,0 +1,166 @@
++++
+title = "Codestitcher: inter-procedural basic block layout optimization"
+extra.bio = """
+Eashan Garg is an undergraduate studying CS and English.
+Gautam Mekkat is an undergraduate studying CS and ECE.
+"""
+extra.author = "Eashan Garg and Gautam Mekkat"
++++
+
+## Introduction
+
+So far, we've covered traditional compiler optimizations such as dead code elimination, copy propagation, and constant propagation. These techniques are centered around *elimination*--- reducing the number of instructions in a program. This is useful, but programs have dramatically increased in size over the years. Eliminating unnecessary code can only do so much. Meanwhile, the capacities of memory units like L1 caches and the TLB haven't grown to account for larger program sizes. One reason for this is that as cache size increases, latency increases. Because of this, we want to optimally place instructions in caches, making sure the most commonly run code lives in the cache. Code layout is critical here. In this post we will discuss this paper's approach to code layout.
```

Comment:
```suggestion
So far, we've covered traditional compiler optimizations such as dead code elimination, copy propagation, and constant propagation. These techniques are centered around *elimination*---reducing the number of instructions in a program. This is useful, but programs have dramatically increased in size over the years. Eliminating unnecessary code can only do so much. Meanwhile, the capacities of memory units like L1 caches and the TLB haven't grown to account for larger program sizes. One reason for this is that as cache size increases, latency increases. Because of this, we want to optimally place instructions in caches, making sure the most commonly run code lives in the cache. Code layout is critical here. In this post we will discuss this paper's approach to code layout.
```

Do not put spaces around em dashes. https://capra.cs.cornell.edu/styleguide/#emdash

Reply:
Because your audience for this post includes "the rest of the world," it's a little odd to say "we" have covered traditional compiler optimizations. Maybe clarify that you're talking about the lectures in the first few weeks of CS 6120? Or just leave out the whole "we" notion altogether?

Reply:
> making sure the most commonly run code lives in the cache.

Not just that's in the cache, but that it's in the cache when it's time to execute it. Namely, that most accesses to load instructions hit in the L1 I-cache.

Reply:
> optimally place instructions in caches

We unfortunately don't get to place anything in caches explicitly. Caches are implicit. All we can do is place things in "memory" (abstractly) nearby to each other and hope that this means the cache treats them well.